src/java_bytecode/java_object_factory.cpp

```diff
@@ -593,21 +598,47 @@ codet initialize_nondet_string_struct(
         code_assumet(binary_relation_exprt(length_expr, ID_le, max_length)));
     }
 
+    // Data array needs to be dynamically allocated, to avoid reusing the
+    // same pointer in loops and function calls.
+    // char (*array_data_init)[INFINITY];
+    // array_data_init = allocate char[INFINITY]
+    symbolt &data_pointer_sym = get_fresh_aux_symbol(
+      pointer_type(
+        array_typet(java_char_type(), infinity_exprt(java_int_type()))),
+      "string_data_pointer",
+      "string_data_pointer",
+      loc,
+      ID_java,
+      symbol_table);
+    const auto data_pointer = data_pointer_sym.symbol_expr();
+    code.add(code_declt(data_pointer));
+
+    side_effect_exprt java_new_array(ID_allocate);
+    java_new_array.copy_to_operands(infinity_exprt(java_int_type()));
+    java_new_array.copy_to_operands(false_exprt());
+    java_new_array.type() = data_pointer.type();
+    code.add(code_assignt(data_pointer, java_new_array));
+
+    // `data_expr` is `*data_pointer`
     // data_expr = nondet(char[INFINITY]) // we use infinity for variable size
-    exprt data_expr = make_nondet_infinite_char_array(symbol_table, loc, code);
+    const dereference_exprt data_expr(data_pointer);
+    const exprt nondet_array =
+      make_nondet_infinite_char_array(symbol_table, loc, code);
+    code.add(code_assignt(data_expr, nondet_array));
```

Comment:
Suggestion: Each of those sections starting with a comment, and ending with a `code.add` could be put into a dedicated function named after the comment. For instance here - `make_assign_nondet_infinite_char_array` or something similar. That way, you could express the intent of a code section with the code itself.